dace/libraries/sparse/nodes/csrmm.py

```diff
@@ -0,0 +1,688 @@
+        else:
+            alpha = constants[node.alpha]
+            beta = constants[node.beta]
+
+        # Set up options for code formatting
+        # opt = _get_codegen_gemm_opts(node, state, sdfg, adesc, bdesc, cdesc, alpha, beta, cdtype, func)
+
+        opt = {}
+
+        opt['arr_prefix'] = arr_prefix = ''
+        if needs_copy:
+            opt['arr_prefix'] = arr_prefix = '_conn'
+
+        opt['func'] = func
+
+        if node.opA == 1:
+            opt['opA'] = 'CUSPARSE_OPERATION_TRANSPOSE'
+        elif node.opA == 2:
+            opt['opA'] = 'CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE'
+        else:
+            opt['opA'] = 'CUSPARSE_OPERATION_NON_TRANSPOSE'
+
+        opt['opB'] = 'CUSPARSE_OPERATION_NON_TRANSPOSE'
+
+        opt['layout'] = 'CUSPARSE_ORDER_ROW'
+
+        opt['compute'] = f'CUDA_R_{to_cublas_computetype(dtype)}'
+        opt['handle'] = '__dace_cusparse_handle'
+
+        opt['alpha'] = alpha
+        opt['beta'] = beta
+
+        opt['nrows'] = cdesc.shape[0]
+        opt['ncols'] = cdesc.shape[1]
+        opt['arows'] = cdesc.shape[0]
+        opt['acols'] = bdesc.shape[0]
+        opt['bcols'] = bdesc.shape[1]
+        opt['annz'] = avals.shape[0]
+        if node.opA != 0:
+            opt['arows'], opt['acols'] = opt['acols'], opt['arows']
+
+        opt['ldb'] = opt['ncols']
+        opt['ldc'] = opt['ncols']
+
+        call = """
+            cusparseSpMatDescr_t matA;
+            cusparseDnMatDescr_t matB, matC;
+            void*                dBuffer    = NULL;
+            size_t               bufferSize = 0;
+            // Create sparse matrix A in CSR format
+            dace::sparse::CheckCusparseError( cusparseCreateCsr(&matA, {arows}, {acols}, {annz},
+                                                {arr_prefix}_a_rows, {arr_prefix}_a_cols, {arr_prefix}_a_vals,
+                                                CUSPARSE_INDEX_32I, CUSPARSE_INDEX_32I,
+                                                CUSPARSE_INDEX_BASE_ZERO, {compute}) );
+            // Create dense matrix B
+            dace::sparse::CheckCusparseError( cusparseCreateDnMat(&matB, {acols}, {bcols}, {ldb}, {arr_prefix}_b,
+                                                {compute}, {layout}) );
+            // Create dense matrix C
+            dace::sparse::CheckCusparseError( cusparseCreateDnMat(&matC, {arows}, {bcols}, {ldc}, {arr_prefix}_c,
+                                                {compute}, {layout}) );
+            // allocate an external buffer if needed
+            dace::sparse::CheckCusparseError( cusparseSpMM_bufferSize(
+                                            {handle},
+                                            {opA},
+                                            {opB},
+                                            {alpha}, matA, matB, {beta}, matC, {compute},
+                                            CUSPARSE_SPMM_ALG_DEFAULT, &bufferSize) );
+            cudaMalloc(&dBuffer, bufferSize);
```

Comment:
I really hate this. I know it was in the original implementation but it's an uncalled-for malloc/free

Reply:
How will you avoid this? I guess we need to use the memory pool, but I haven't looked at the API for that.